cdap-app-templates/cdap-program-report/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala

```diff
@@ -0,0 +1,185 @@
+    * |   run   |  record  |  required columns, filter columns, sort columns ...
+    * +---------+----------+-----------------------------------------------------
+    * For instance, if the required columns, filter columns and sort columns combined only contain three columns
+    * "namespace", "program", and "duration", the aggregated [[org.apache.spark.sql.DataFrame]]
+    * will contain columns as shown below:
+    * +---------+----------+---------------+-----------+------------+
+    * |   run   |  record  |   namespace   |  program  |  duration  |
+    * +---------+----------+---------------+-----------+------------+
+    * After filtering and sorting are done on the [[org.apache.spark.sql.DataFrame]],
+    * only the columns required in the report will be kept in the [[org.apache.spark.sql.DataFrame]] as shown below:
+    * +---------------------+--------------------+----------------------
+    * |  required column 1  | required column 2  | required columns ...
+    * +---------------------+--------------------+----------------------
+    * For instance, if the required columns only contain three columns "namespace", "program", and "run",
+    * the final [[org.apache.spark.sql.DataFrame]] will contain columns as shown below:
+    * +---------+---------------+-----------+
+    * |   run   |   namespace   |  program  |
+    * +---------+---------------+-----------+
+    * The final [[org.apache.spark.sql.DataFrame]] will be written to a JSON file at the given output location,
+    * accompanied by an empty _SUCCESS file indicating success.
+    *
+    * @param spark the spark session to run report generation with
+    * @param request the request containing the requirement for the report generation
+    * @param inputURIs URIs of the avro files containing program run meta records
+    * @param outputLocation location of the output directory where the report file and _SUCCESS file will be written
+    * @throws java.io.IOException when fails to write to the _SUCCESS file
+    */
+  @throws(classOf[IOException])
+  def generateReport(spark: SparkSession, request: ReportGenerationRequest, inputURIs: java.util.List[String],
+                     outputLocation: Location): Unit = {
+    import spark.implicits._
+    val df = spark.read.format("com.databricks.spark.avro").load(inputURIs: _*)
+    // Group the program run meta records by program run Id's and aggregate grouped records into a column
+    // with data type Record. The aggregated DataFrame aggDf will have two columns: "run" and "record"
+    val aggCol = new RecordAggregator().toColumn.alias(RECORD_COL).as[Record]
+    // TODO: configure partitions. The default number of partitions is 200
+    var aggDf = df.groupBy(Constants.RUN).agg(aggCol)
+    // Construct a set of fields to be included in the final report with required fields and fields from the request
+    val reportFields: collection.mutable.LinkedHashSet[String] =
```

Comment:
You don't need a mutable set.

Reply:
Just do:

```scala
// Also change REQUIRED_FIELDS to a Set instead of Seq
val reportFields = REQUIRED_FIELDS ++ Option(request.getFields).map(_.toSet).getOrElse(Nil)
```